src/pubsub/ua_pubsub_ns0.c

```diff
@@ -1849,6 +1849,149 @@ setSecurityKeysAction(UA_Server *server, const UA_NodeId *sessionId, void *sessi
+    UA_UInt32 requestedKeyCount = *(UA_UInt32 *)input[2].data;
+
+    UA_PubSubKeyStorage *ks =
+        UA_Server_findKeyStorage(server, *securityGroupId);
+    if(!ks)
+        return UA_STATUSCODE_BADNOTFOUND;
+
+    UA_Boolean executable = UA_FALSE;
+    UA_SecurityGroup *sg = UA_SecurityGroup_findSGbyName(server, *securityGroupId);
+    void *sgNodeCtx;
+    getNodeContext(server, sg->securityGroupNodeId, (void **)&sgNodeCtx);
+    executable = server->config.accessControl.getUserExecutableOnObject(
+        server, &server->config.accessControl, sessionId, sessionHandle, methodId,
+        methodContext, &sg->securityGroupNodeId, sgNodeCtx);
+
+    if(!executable)
+        return UA_STATUSCODE_BADUSERACCESSDENIED;
+
+    /* If the caller requests a number larger than the Security Key Service permits, then
+     * the SKS shall return the maximum it allows.*/
+    if(requestedKeyCount > sg->config.maxFutureKeyCount)
+        requestedKeyCount =(UA_UInt32) sg->keyStorage->keyListSize;
+    else
+        requestedKeyCount = requestedKeyCount + currentKeyCount; /* Add Current keyCount */
+
+    /*The current token is requested by passing 0.*/
+    UA_PubSubKeyListItem *startingItem = NULL;
+    if(startingTokenId == 0) {
+        startingItem = sg->keyStorage->currentItem;
+    } else {
+        retval = UA_PubSubKeyStorage_getKeyByKeyID(
+            startingTokenId, sg->keyStorage, &startingItem);
+        /*If the StartingTokenId is unknown, the oldest (firstItem) available tokens are
+         * returned. */
+        if(retval == UA_STATUSCODE_BADNOTFOUND)
+            startingItem =  TAILQ_FIRST(&sg->keyStorage->keyList);
+    }
+
+    /*SecurityPolicyUri*/
+    retval = UA_Variant_setScalarCopy(&output[0], &(sg->keyStorage->policy->policyUri),
+                         &UA_TYPES[UA_TYPES_STRING]);
+    if(retval != UA_STATUSCODE_GOOD)
+        return retval;
+
+    /*FirstTokenId*/
+    retval = UA_Variant_setScalarCopy(&output[1], &startingItem->keyID,
+                                      &UA_TYPES[UA_TYPES_INTEGERID]);
+    if(retval != UA_STATUSCODE_GOOD)
+        return retval;
+
+    /*TimeToNextKey*/
+    UA_DateTime baseTime = sg->baseTime;
+    UA_DateTime currentTime = UA_DateTime_nowMonotonic();
+    UA_Duration interval = sg->config.keyLifeTime;
+    UA_Duration timeToNextKey =
+        (UA_Duration)((currentTime - baseTime) / UA_DATETIME_MSEC);
+    timeToNextKey = interval - timeToNextKey;
+    retval = UA_Variant_setScalarCopy(&output[3], &timeToNextKey,
+                                      &UA_TYPES[UA_TYPES_DURATION]);
+    if(retval != UA_STATUSCODE_GOOD)
+        return retval;
+
+    /*KeyLifeTime*/
+    retval = UA_Variant_setScalarCopy(&output[4], &sg->config.keyLifeTime,
+                         &UA_TYPES[UA_TYPES_DURATION]);
+    if(retval != UA_STATUSCODE_GOOD)
+        return retval;
+
+    /*Keys*/
+    UA_PubSubKeyListItem *iterator = startingItem;
```

Comment:
Creating the array and then freeing it, that's a bit too much overhead.
What you can do instead is calloc the array directly into output[2].
Just have a look at UA_Variant_setArray how to set up an array variant.
And then copy over piecewise the elements.
If an element copy fails you can immediately drop out.
And output[2] will be cleaned up by the stack after sending the response.
The not-yet-set elements will be zeroed (initialized) already.